_published/0013-authentication.md

```diff
@@ -0,0 +1,165 @@
+validity and authenticity of the token. See
+[validating an OIDC ID-Token](https://connect2id.com/blog/how-to-validate-an-openid-connect-id-token).
+
+#### Sessions
+The wharf-api is currently stateless and does not use sessions. It is
+possible to either include the ID token in every request to the server or to
+start using sessions to store the ID token between requests.
+
+#### Local development vs production environments
+For the initial release of the feature the configuration will be tailored to
+using azure as an IDP. These settings will as necessary be exposed as
+environment variables to each respective docker image.
+
+
+#### Choice of Client
+
+The OIDC standard defines a
+[list of certified clients](https://openid.net/developers/certified/)
+that can help simplify the implementation of ID Token retrieval.
+
+For the backend a full client will not be required. As the tokens just need
+to be validated and not requested. Validation can be done offline by caching
+the issuers' signature for future validations (OBS! The signature key may
+rotate on a timely basis).
+
+## Compatibility
+
+Expect no compatability between versions of wharf without authentication to
+versions of wharf with authentication. There is a potential threat in leaving
+the system backwards compatible in that a non-secured version of the api could
+run together in an otherwise secure cluster.
+
+We should aim to pick a solution where any IDP supporting OIDC and RBAC can be
+used. Locking ourselves to a single IDP when a standard such as OIDC exists
+seems like a  waste opportunity. Keeping the choice of IDP open will leave
+more room for choice and flexibility in the future.
+
+## Alternative solutions
+
+There are many types of identity verification systems.
+This RFC assumes that OIDC will be used. Within OIDC there are multiple
+different [flow definitions](https://developer.okta.com/docs/concepts/oauth-openid/#choosing-an-oauth-2-0-flow).
+We should primarily use `codeflow with pkce` as this is presumed to be the
+most secure, and best compatible at the time of writing.
+
+## Future possibilities
+
+We may want to implement multiple different IDP sources. This could allow
+login from different sources than the Iver-Azure-AD. A sample situation
+where this could be useful is where/when we wish to give a client
+view/control over the deployment pipeline.
+
+When handling ID-Tokens from RBAC enabled IDPs its is possible to verify
+custom permission levels using custom role claims.
+[See this link](https://docs.microsoft.com/en-us/azure/role-based-access-control/custom-roles)
+for more information about custom claims. This system allows for setup
+fine-tuned access control on a per-deployment basis. The initial setup
+for each of the different RBACs will likely require setup scripts on a
+per-provider basis as it requires interaction directly with each respective
+cloud on a non-standard basis.
```

Comment:
nit: Would rather see this just documented instead of scripted. It should be such a one-time and only necessary when users needs more access. And just having some example docs of how to properly set this up with different IDPs so those with undocumented IDPs could connect the dots themselves.

Don't know exactly what you refer to with "setup scripts", so excuse me if I misunderstood

Reply:
I have reworded this such that its left up to the implementation. 

The decision on which IDP to use for authentication will need to be made and customised in setup per instance.

Since this is not a one-time setup i would personally have this setup be automated rather than manual. Personally i also belive code can serve as documentation. The setup is likely needed to be done one per instance of wharf.

Reply:
Alright! Convinced :)